flax/linen/module.py

```diff
@@ -1011,60 +1050,144 @@ def __call__(self, train: Optional[bool] = None):
     return a
 
 
-  # THE PART BELOW IS STILL UNDER DEVELOPMENT, PLEASE IGNORE
-  # ===========================================================
-  # @contextmanager
-  # def mutate(self, mutable=True, **updates):
-  #   cloned = self.clone(**updates)
-  #   try:
-  #     cloned.scope._variables = _unfreeze_variables(
-  #         cloned.scope._variables, mutable)
-  #     yield cloned
-  #   finally:
-  #     cloned.scope._variables = freeze(cloned.scope._variables)
-
-  # def initialized(self, rngs, *args, method='__call__', **kwargs):
-  #   if self.parent is not None:
-  #     raise ValueError("Pattern for initialized is "
-  #                      "`Module(parent=None, ...attrs...).initialized(...)`")
-  #   scope = Scope(variables={}, rngs=rngs)
-  #   with self.mutate(parent=scope) as initialized:
-  #     if method is not None:
-  #       getattr(initialized, method)(*args, **kwargs)
-  #   return initialized
-
-  # @property
-  # def variables(self):
-  #   """Get a view of Module variables with easy dot-syntax navigation."""
-  #   return DotGetter(self.scope.variables())
-
-  # def __getattr__(self, name):
-  #   # Used for easy colab/jupyter introspection, and to provide a
-  #   # consistent top-level interface to self.<attr> for both simple
-  #   # and multi-method modules.
-  #   if name in self.children:
-  #     val = self.children[name]
-  #     if isinstance(val, str):  # variable
-  #       return self.variables[val][name]
-  #     else:  # submodule
-  #       val.scope = self.scope.push(name)
-  #       self.scope.reservations.remove(name)
-  #       return val
-  #   else:
-  #     raise AttributeError(
-  #         f"'{self.__class__.__name__}' object has no attribute '{name}'")
-
-  # def __dir__(self):
-  #   return list(self.children.keys()) + object.__dir__(self)
-
-  # TODO: Should this be what `clone` always does if you don't pass in an explicit
-  # parent?
-  # def detached(self):
-  #   return self.clone(parent=None)
-
-  # # TODO: Consider whether this is a helpful abstraction, and think about naming.
-  # # See its use in design_test/linen/weight_std.py
-  # def materialized(self, variables={}, rngs={}):
-  #   assert self.scope is None, ("Can't attach a module twice."
-  #                               " Maybe you want to clone first?")
-  #   return self.clone(parent=Scope(variables, rngs))
+def apply(fn: Callable[..., Any], module: Module,
+          mutable: CollectionFilter = False,
+          capture_intermediates: Union[bool, Callable[[Module, str], bool]] = False) -> Callable[..., Any]:
+  """Creates an apply function for the given function and ``Module`` instance.
```

Comment:
What about

```
  """Creates an apply function to call ``fn`` with a bound module.
```

I feel mentioning `bound module` makes it clearer what this function is for...
(same could be argued for other 1st lines pydoc strings)

And/or a reference to `Module.bind()` would help understanding this function.